core: Adding CUDA IOV copy helpers

rajachan opened the pull request:
~~This PR is built on https://github.com/ofiwg/libfabric/pull/5561, so ignore the rdma-core related commits and just look at the two top commits in this branch. Github doesn't handle parent-child PR relationships well across forks.~~ (The latest version has the right commits.)

@shefty @j-xiong I have added the copy functions to EFA in this case, but I see no reason why they can not be general enough and live in a ofi_cuda.h. I know we have talked about a generic peer software interface for providers that support `FI_HMEM` in the past, so until that gets defined I did not want to push this to core. Let me know what you think.

I have more changes to support `FI_HMEM` in the EFA provider lined up, but they depend on these helpers (and the parent PR), so I'll wait for these two to get merged before cutting more PRs.

shefty commented:
Looking at just the cuda support, I would go ahead and put them into an ofi_cuda.h file.  Even if we need to change them in the future, at least we're starting with the same set of calls, so everyone gets updated.

shefty commented on prov/efa/src/efa_fabric.c:
```diff
@@ -853,6 +854,29 @@ int efa_fabric(struct fi_fabric_attr *attr, struct fid_fabric **fabric_fid,
+	if (ret)
+		return -errno;
+
+	ret = ibv_fork_init();
```
FWIW, we had this call in the verbs provider, but had to remove it.  It caused a performance hit, but also didn't work.  An application might access verbs through multiple middleware libraries, and there's no guarantee that the other library doesn't do something with verbs prior to this being called.  The only viable solution was to use the verbs fork environment variable.  Once that's a requirement, there's no need to call this, which also removes the performance impact when fork isn't used.

rwespetal replied:
Right, this isn't ideal from both a performance standpoint and avoiding a race with something accessing verbs before Libfabric initialization. However, we're concerned that our customers are not going to know whether they need fork support or not. We should be able to mostly avoid another middleware library accessing verbs on EC2. I put the call in EFA's fabric init as to avoid impacting other Libfabric providers here with our workaround.

jswaro replied:
This caused a problem with initialization that was a functional problem, not a performance problem. I'll see if I can dig up the issue.

jswaro replied:
See https://github.com/ofiwg/libfabric/pull/4973

shefty commented on prov/efa/configure.m4:
```diff
@@ -52,23 +52,28 @@ AC_DEFUN([FI_EFA_CONFIGURE],[
+						 [CUDA support])],
+				      [efa_happy=0],
+				      [])],
+		    [])
```
I'm not sure if this is the best place to check for cuda availability.  @aingerson - any opinions on this?

rajachan replied:
If I moved it to ofi_cuda.h, I was planning to add it to the main configure.ac.

aingerson replied:
Yeah I would put this check in the main configure and define a global HAVE_LIBCUDA for all providers to use.
I would also lean towards removing the --with-cuda option and just checking for the presence of the libraries/functions and enable HAVE_LIBCUDA if they are present.

rajachan replied:
Hmm, not sure we'd want that. A user might want to have a build that doesn't link to CUDA runtime libraries on a system that does have all the dependencies. Having an explicit flag is better, IMO.  

This also makes it convenient to provide custom paths (`--with-cuda=/usr/local/cuda/`)

shefty replied:
Users can already provide custom paths to libraries.  I don't like forcing users to know what path a library was installed into.  If we want an option to disable the cuda checks, that would be fine, but I would not go the route of making it difficult in the common use case.

aingerson replied:
Well the user would still have to add the paths to the CFLAGS (I've only ever seen it in /usr/local/cuda) so they could choose to omit it/provide a custom path.
I'll defer to Sean on which option would be best.

aingerson replied:
Sent it before reading Sean's response. I agree with him. If you want to add a --disable-cuda option then I think that's the better option.

rajachan replied:
Yeah, CUDA installs to /usr/local/cuda/ by default and setting CFLAGS/LDFLAGS by hand is worse than giving a configure flag. I'd still prefer doing a AC_ARG_WITH (which can do a without). I'll leave it enabled by default to get the behavior you want.

rajachan commented:
> Looking at just the cuda support, I would go ahead and put them into an ofi_cuda.h file. Even if we need to change them in the future, at least we're starting with the same set of calls, so everyone gets updated.

Alright then, will move it over to common code.

iziemba commented:
Having providers integrate with the device memory interfaces seems like a bottleneck in the design for supporting new devices. Instead of explicitly calling out a device memory interface, has a generic, callback-based approach been consider?

Applications define the callbacks to be used for a specific device memory region. For example, instead of having CUDA calls integrated into libfabric, applications define a callback structure which is passed to the provided via `fi_mr_regattr().` This callback structure defines functions the provider uses to copy to host from device and vice versa. The following is an example how `struct fi_mr_attr` could be updated to accommodate this.
```
enum fi_hmem_iface {
	FI_HMEM_SYSTEM	= 0,
	FI_HMEM_DEVICE,
};

struct fi_ops_mr_hmem {
	void	(*memcpy_htod)(void *dev_buf, void *host_buf, size_t n);
	void	(*memcpy_dtoh)(void *host_buf, void *dev_buf, size_t n);
};

struct fi_mr_attr {
	const struct iovec	*mr_iov;
	size_t			iov_count;
	uint64_t		access;
	uint64_t		offset;
	uint64_t		requested_key;
	void			*context;
	size_t			auth_key_size;
	uint8_t			*auth_key;
	enum fi_hmem_iface	iface; // If FI_HMEM_DEVICE, hmem_ops pointer is valid.
	struct fi_ops_mr_hmem	*hmem_ops;
};
```
Thoughts?

rajachan commented:
I don't know if we should let applications register device-specific callbacks, that's just punting the problem to a different layer, and each application/middleware using libfabric would have to provide callbacks for all devices they wish to support.

The generic peer device interface I was alluding to was something OFIWG talked about in the past, when adding the FI_HMEM MR bits. We would need libfabric core to abstract device-specific ops (data movement or otherwise) that providers can use use to support any device. Application just has to pass down information about the type of memory and providers can use the abstracted core functions without writing device-specific code for each new device they want to support.

The EFA patch I have here is just getting ahead of that, in the interest of adding CUDA buffer support sooner rather than later. Once the core interfaces are defined, we intend to migrate to that.

shefty commented:
Pushing the responsibility for implementing copies from host to device into every application is not the direction I would choose to go.  That requires even more work to support new devices than having a centralized library handle this.

The data movement is not limited, for example, between GPU and host memory.  The data movement may be directly to/from the GPU and the fabric.  The fabric provider needs to have the details necessary to setup such transfers, such as configuring PCI peer to peer.

shefty commented:
Adding to @rajachan comment, internally it's likely that we'll end up with a solution that sets up and uses callbacks from the providers into libfabric core code once we have support for multiple devices.  So, there's a good chance we'll have callbacks, just not passed in by the application.

iziemba commented:
> Pushing the responsibility for implementing copies from host to device into every application is not the direction I would choose to go. That requires even more work to support new devices than having a centralized library handle this.

I'm not sure I follow on how this would be more work. The additional work is placed in the application layer which has the best knowledge for how libfabric would interface with a device. If the callbacks were well-defined, libfabric support for new devices should be a noop.

> The data movement is not limited, for example, between GPU and host memory. The data movement may be directly to/from the GPU and the fabric. The fabric provider needs to have the details necessary to setup such transfers, such as configuring PCI peer to peer.

By configuring PCI peer to peer, are you referring to setting up the device memory DMA mappings used by the NIC? If so, this seems to be a function of the NIC driver instead of the provider.

shefty commented:
Asking OpenMPI, Intel MPI, MPICH, MVAPICH, GasNET, Charm++, OpenSHMEM, etc. to add device support, and then asking them to all update when there's a new device is far more work than isolating the changes into libfabric.

The kernel driver is an essential component of most libfabric providers. GPU support requires changes to both for ideal performance.

iziemba commented:
> Asking OpenMPI, Intel MPI, MPICH, MVAPICH, GasNET, Charm++, OpenSHMEM, etc. to add device support, and then asking them to all update when there's a new device is far more work than isolating the changes into libfabric.

For clarity, are not all these going to have to update when libfabric defines a new iface enum value (assuming the library is using the new device interface)?

shefty commented:
Not necessarily.  A hooking provider can be be configured and installed at run time.  That provider will check if an address belongs to host or device memory and set the appropriate flags for the core providers.

For best performance, knowledge of device memory should be passed through the stack.  However, some middleware (MPI) loses that information through its API and must relearn it.  The becomes the role of the hooking provider here.

libfabric will detect at installation time what device libraries are available.  If there's only one, that makes life easy.  If there is more than one, an administrator can configure which device library to use, or libfabric will need to pass the address to all of them to see which one claims it.

The hooking capability is really a short term band-aid.  Somewhere higher up in the stack the code knows exactly which device is being used and is calling a third party API to invoke kernels to perform some computation.  The long term goal is to convey that data through the stack.

The API changes are around the long term objective.  Short term, the intent is that core providers can rely on the API, and a hooking provider above them will provide the necessary intercepts and checks.

At a low level, the performance of data transfers between devices and between the host and the device will depend heavily on the device implementation.  For example, some device/platforms will perform much better if transfers are performed as writes from the device, rather than reads across PCI from the host or another device.  Similarly, some NICs will support PCI peer-to-peer transfers, while others may not.  I believe this level of optimization is better off pushed closer to the devices being used.

iziemba commented:
Just to verify if I understood correctly. Two approaches are being taken: short term with a hook provider and long term with ifaces. If the hook provider is installed, applications do not have to distinguish between device/host memory since the hook will perform pointer checks and dynamically select the iface. Does this mean that applications do not have to use `fi_mr_regattr()` to register device memory?

shefty commented:
Correct.  Upper middleware would not use the new registration function to register device memory if hooking is enabled.  There are some obvious drawbacks here.  Addresses used for outbound transfers would be checked on each call, some caching mechanisms would need to be employed, and the device must be supporting a unified virtual address space.  My guess is that the hooking implementation would only support one device interface at a time, but that's unlikely to have an impact in practice.

jswaro commented:
> Pushing the responsibility for implementing copies from host to device into every application is not the direction I would choose to go. That requires even more work to support new devices than having a centralized library handle this.

A hook provider would require knowledge of every device that the provider needs to interact with. I think that this ends up being provider that never finishes development, because we need constantly add new hooks for new device APIs. Today, it's just Cuda. Tomorrow, maybe ROCm. Some time later, a different device API. Add another dimension in the form of API versions. Almost a 'libfabric' within libfabric. ISV codes which might be built against a different version of the cuda library than libfabric might have different behaviors or even different API definitions. It might be a can of worms that can't be easily closed again. 

Libfabric needs to support interactions with device memory, but I think there could be a broader discussion on the topic.

shefty commented:
We're primarly talking about using the libfabric API to transfer data between device memory and the fabric.  There are any number of device API calls that a provider must make to achieve this.  For performance reasons, it may be better to proxy the data through host memory.  That doesn't mean that these are the only device API calls that the provider must make.  There are any number of calls that might be needed, depending on both the provider and device.

The providers need to know of what third party interfaces are being used to access the device.  That's not the same as needing to program to every device.  The libfabric API is designed to handle this, so that it's easy to add support for CUDA, ROCm, OneAPI, etc.

It's not feasible to define all the callbacks that a given provider will need.  Yes, if new device APIs are endlessly defined, the hooking provider would constantly need to update.  So would *every* application.  This is exactly why this functionality needs to be pushed into *one* location.

iziemba commented:
> It's not feasible to define all the callbacks that a given provider will need. Yes, if new device APIs are endlessly defined, the hooking provider would constantly need to update. So would _every_ application. This is exactly why this functionality needs to be pushed into _one_ location.

In addition to being updated for new device interfaces, would not the hooking provider potentially have to be updated to a support a new version of a device library?

shefty commented:
If a device library breaks backwards compatibility with itself, then every software that uses it would need to be updated, which would include libfabric.  If a device library maintains compatibility, but provides additional APIs that could improve performance, then all software that wants the new API would need to change, which too includes libfabric.

These arguments strengthen the need to have this logic in one location, rather than requiring all applications to change.  It also makes it less likely that we could define callback APIs that also wouldn't need to change or be expanded on.

The alternative being debated here is to disallow any provider from providing optimized data transfers to/from device memory.  It would be great if there were standard linux system calls that a provider could make to setup these sort of transfers, but those currently don't exist, and even if/when they do, it would take years before they would be available on most customer systems.

naveen-rn commented:
> There are any number of calls that might be needed, depending on both the provider and device.

@shefty Does this mean that each provider should support the device-specific functions inside the provider level and not introduce a generic interface that can be used by all the providers? 

I'm confused with the expected level of abstraction in the hook provider you are describing. AFAIU, there are two possible levels:
1. per-provider
2. per-device

per-provider would allow each provider to define its own list of device APIs and can be used only by the provider. For example, as it is in this PR: efa_cuda.h, efa_rocm.h and so on.

per-device would define a list of generic device APIs that is common to all providers. For example, we would have something like ofi_cuda.h, ofi_rocm.h and so on.

If we could come-up with the per-device option - I'm failing to understand why it can't be made as a separate library maintained by the community or proprietary? Is there a reason, why it should be tightly coupled with the libfabric core? Are we expecting to make use of some other resource from the libfaric-core by that hook? Please clarify.

jswaro commented:
> The providers need to know of what third party interfaces are being used to access the device.

Agreed.  

> The libfabric API is designed to handle this, so that it's easy to add support for CUDA, ROCm, OneAPI, etc.

I agree that there is merit in facilitating copy behavior but I'd like to discuss options in how libfabric's API approaches the problem. It has been years since I've written a cuda application, but it used to be very expensive to copy to/from the device. I don't think facilitating data transfer for certain libfabric capabilities such as inject is going to improve performance. If it has gotten better, then maybe it makes sense to allow all data transfers from device to host. Will all providers need to write device-specific code to handle device data transfer since we are pushing it down?

> So would _every_ application. This is exactly why this functionality needs to be pushed into _one_ location.

Understood. Someone will have to maintain the code that is used to perform actions on device memory whether that action is a memory hook, or a copy. I'd like to have discussion on what is the best approach for device memory interactions with providers. 

It is possible that the best solution for API compatibility with libraries isn't provided by libfabric, **but it might be**. In libfabric, we **can** provide a centralized solution to device-to-host/host-to-device memory transfer -- at a cost of supporting each API involved (which is a small list today). However, there is a cost in terms of memory registration caching, memory-type detection, data transfer overhead, and others which affect libfabric correctness and performance that should be discussed. How would modifications to the cache implementation work with multiple memory monitors, if that is possible? Perhaps device memory cannot be used with the cache. Is scalable memory registration affected by the use of device memory? 

> If a device library breaks backwards compatibility with itself, then every software that uses it would need to be updated, which would include libfabric.

Not necessarily. If a generic callback format was supported by libfabric, and upper layers provided the wrapper around device memory transfer functions as a function pointer, then libfabric does not need to be updated. The update/recompile only applies to upper layers who are the consumers of device memory. The user application needs to recompiled regardless, since they are allocating, performing operations on, and freeing the region. 

Libfabric only needs to be recompiled if we directly associate libfabric with the memory management of device memory. 

Yes, that would force middle-ware to define function wrappers for device memory. If I were a middle-ware writer, I wouldn't be thrilled about it. I am going to point out that middle-ware already needs to be aware that it is device memory, and they probably already know how to interact with it. 

--------------

I'm not advocating for any approach over the other. I'm looking at options and asking why solution X over solution Y. 

> The alternative being debated here is to disallow any provider from providing optimized data transfers to/from device memory.

That's not what I'm getting from this. I see this as a discussion about how libfabric should approach optimized data transfer to/from the device and how that process should work from the application to the provider. If providing a centralized solution in Libfabric is the solution, then so be it. I'm asking about the motivations, what was considered, why it was accepted or discarded... I think this could be a good discussion for the next OFIWG meeting. 

If the best solution is in libfabric -- great. If the best solution is not in libfabric, great.

shefty commented:
This PR only modifies the EFA provider.  It does not change the API at all.  There are a couple of functions which can move out of EFA into common code.  That's the extent of this PR and the changes requested.

If there is a specific proposal to change the API, submit a PR with those changes or bring it up in the ofiwg for discussion.  The approach taken so far aligns with all previous discussions and the current API.

rajachan commented:
First off, two house-keeping notes: I moved the util functions out of EFA into common code and updated configure to reflect that. Also renamed the PR to reflect this and the discussion. Demonstrating the use of this code from within a provider would be useful to this discussion, but I did not want to spend cycles reworking the patch series I have until this gets merged.

---

Looks like I missed some good discussion on Friday. A lot of it has been covered already, so I'll just jump in and comment on some things that stand out:

>  I'm asking about the motivations, what was considered, why it was accepted or discarded... I think this could be a good discussion for the next OFIWG meeting.

Can't agree more, the datapath abstraction to deal with different devices needs more discussion in a regular OFIWG call or f2f (possibly at the OFA workshop, if enough people with interest in the topic are going to be there). FWIW, this was briefly discussed in past OFIWG meetings when we defined the `FI_HMEM` API.  The agreement back then was to get the bare-minimum API needed to enable HMEM/CUDA support (the cap bit, mode bit, the MR attribute and interface type enums, see https://github.com/ofiwg/libfabric/pull/5262) and later give some due diligence to the various devices we might want to support in libfabric and define a southbound provider-facing interface that works for them all. We're knocking on the second item with the discussion here.

> per-provider would allow each provider to define its own list of device APIs and can be used only by the provider. For example, as it is in this PR: efa_cuda.h, efa_rocm.h and so on.

> per-device would define a list of generic device APIs that is common to all providers. For example, we would have something like ofi_cuda.h, ofi_rocm.h and so on.

> completely-generic interface would a common access to a particular functionality and it is upto the hook to make the device selection.

No.1 was not the goal. LIke I mentioned in the original comment, it was an attempt to get ahead of the eventuality, which was no.2. I've updated the PR now so it looks more like no.2. Option no. 2 would also allow for optimizations at a single place. The IOV copy mechanism I've added is pretty rudimentary. Like I called out in the commit message, a low-hanging fruit would be to use gdrcopy to optimize transfers in one direction. I'm sure we can optimize them over time in a manner that benefits all providers. Option no. 3 would involve expensive (possibly multiple) device queries, which we might not want in the criticial send path. We can not avoid this when using a hook provider with a naive app that can not deal with accelators/compute offload devices, but that's the price to pay to prevent putting the onus on end application developers who develop codes on top of MPI/SHMEM/NCCL, etc.

More broadly (a.k.a, rambling alert), we ( theOFI community) are going to have to work with some common libfabric-enabled middleware to give us information about the device that a given buffer belongs to so we can optimize away the redundant buffer detection. Ideally, in a send path, we should be able to directly take an application-provided device buffer and post a work request to the NIC (if the NIC can interface with peer devices) or have the provider copy device buffers to some internal host-based bounce buffer (utility functions for which we can provide in libfabric core, like this PR). Again, thinking of CUDA as an example for a second, most MPI libraries are CUDA-aware today. Without looking too much into all of them, I'll say that will be relatively low-touch to get the MPI implementors to register CUDA buffers differently than how they register host buffers today. Once we have a handful of FI_HMEM provider implementations, I'm sure other middleware developers will show a natural interest too. NCCL's libfabric plugin also knows if an incoming buffer backed a host/device. It will be trivial to extend it to register CUDA buffers with FI_HMEM if a provider's capability allows it. To Sean's point, making a case to introduce this change will be easier than asking middleware developers to implement callbacks for the data movement functions. Of course, a several of us are also involved in higher-level middleware implementations, so it will be good to bring different perspectives to that OFIWG discussion, whenever we have it.


Back to this PR, even if a vendor's NIC can directly DMA data into a peer device, depending on the various protocols a provider for that NIC might implement (eager/segmentations/rendezvous, etc), there might still be a need for the utilities here. Like the example James brought up, you might not want to submit an inject request with a CUDA buffer, or might have to deal with unexpected eager messages for FI_TAGGED, in which case you can not prepost device recv buffers.
